### PR TITLE
Code Generator for Known String Lookup

### PR DESCRIPTION
**Pull request by bchong95**
# Code Generator for Known String Lookup

## Description

This PR adds code to generate code that is able to efficiently lookup if a string belongs to a known set of values. The general idea is that we segment the strings based on the length. If two (or more) strings have the same length, then we differentiate them by doing in an if ladder of mem-compares (which is vectorized). 

For parsing BackendMetrics (end to end) the perf numbers are as follows:

* MemCompare Ladder = 662
* Prefix Tree = 800
* Segmented MemCompare Ladder = 397

Which means improving the tokenizer brought down the end to end time by about 40%. This means the tokenizer itself became much faster. Note the BackendMetrics was already optimized before this PR, so overall this is much faster than the v1 parser.

This approach can also be used for HttpHeaders and similar scenarios.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/known_string_lookup_generator.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿def variable_name(str):
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/known_string_lookup_generator.py]: Can we add a read me or at least comments explaining what this is used for.

File: Microsoft.Azure.Cosmos/src/known_string_lookup_generator.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+token_name = "BackendMetrics"
+
+print(strings)
+
+d = {}
+for string in strings:
+    d.setdefault(len(string), []).append(string)
+
+print(d)
+print(f"public static class {token_name}TokenLookupTable")
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/known_string_lookup_generator.py]: I don't see any class with this name. Am I missing something? Maybe rename it to make it obvious it's auto generated and maybe add a comment as part of the output.

Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/known_string_lookup_generator.py]: It's not always meant to fully generate the class. It generates large pieces that are useful.

File: Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.SystemStrings.cs [hunk 3]
```diff
@@ -53,26 +53,292 @@ private static class SystemStrings
 
             public static readonly ReadOnlyMemory<byte>[] Utf8Values = Utf16Values.Select(x => (ReadOnlyMemory<byte>)Encoding.UTF8.GetBytes(x)).ToArray();
 
-            public sealed class Dictionary
+            public static class TokenBuffers
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Json/JsonBinaryEncoding.SystemStrings.cs]: Please add comments where the script is used to generate the content.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Query/Metrics/Performance.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Microsoft.Azure.Cosmos.Tests.Query.Metrics
+{
+    using System;
+    using VisualStudio.TestTools.UnitTesting;
+    using Microsoft.Azure.Cosmos.Query.Core.Metrics;
+    using System.Diagnostics;
+
+    [TestClass]
+    public class Performance
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Query/Metrics/Performance.cs]: Performance test should be added to the Performance tests 

C:\azure-cosmos-dotnet-v3\Microsoft.Azure.Cosmos\tests\Microsoft.Azure.Cosmos.Performance.Tests

**Pull request merged**